kernel/rtlil.cc

```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
-		bits.push_back(bit);
+		bv.push_back(bit);
 }
 
 RTLIL::Const::Const(const std::vector<bool> &bits)
 {
 	flags = RTLIL::CONST_FLAG_NONE;
-	this->bits.reserve(bits.size());
+	backing = bitvectype();
+	bitvectype& bv = assert_get_bits("Const::Const(std::vector<bool>)");
+	bv.reserve(bits.size());
 	for (const auto &b : bits)
-		this->bits.emplace_back(b ? State::S1 : State::S0);
+		bv.emplace_back(b ? State::S1 : State::S0);
+}
+
+[[nodiscard]] RTLIL::Const::bitvectype& RTLIL::Const::assert_get_bits(const char* ctx) const {
+	// return assert_get<bitvectype, decltype(backing)>(&backing, ctx);
+	return std::get<bitvectype>(backing);
+}
+
+[[nodiscard]] std::string& RTLIL::Const::assert_get_str(const char* ctx) const {
+	// return assert_get<std::string, decltype(backing)>(&backing, ctx);
+	return std::get<std::string>(backing);
 }
 
 bool RTLIL::Const::operator <(const RTLIL::Const &other) const
 {
-	if (bits.size() != other.bits.size())
-		return bits.size() < other.bits.size();
-	for (size_t i = 0; i < bits.size(); i++)
-		if (bits[i] != other.bits[i])
-			return bits[i] < other.bits[i];
+	const char* ctx = "operator<";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
+		return decode_string() < other.decode_string();
+
+	if (std::get_if<std::string>(&backing))
+		return assert_get_str(ctx) < other.assert_get_str(ctx);
+
+	bitvectype& bv = assert_get_bits(ctx);
+	auto other_bv = other.assert_get_bits(ctx);
+
+	if (bv.size() != other_bv.size())
+		return bv.size() < other_bv.size();
+
+	for (size_t i = 0; i < bv.size(); i++)
+		if (bv[i] != other_bv[i])
+			return bv[i] < other_bv[i];
 	return false;
 }
 
 bool RTLIL::Const::operator ==(const RTLIL::Const &other) const
 {
-	return bits == other.bits;
+	const char* ctx = "operator==";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
+		return decode_string() == other.decode_string();
+
+	if (std::get_if<std::string>(&backing))
+		return assert_get_str(ctx) == other.assert_get_str(ctx);
+
+	return assert_get_bits(ctx) == other.assert_get_bits(ctx);
 }
 
 bool RTLIL::Const::operator !=(const RTLIL::Const &other) const
 {
-	return bits != other.bits;
+	return !(*this == other);
+}
+
+std::vector<RTLIL::State>& RTLIL::Const::bits()
+{
+	bitvectorize();
+	return assert_get_bits("Const::bits()");
+}
+
+const std::vector<RTLIL::State>& RTLIL::Const::bits() const
+{
+	bitvectorize();
+	return assert_get_bits("Const::bits()");
+}
+
+
+std::vector<RTLIL::State> RTLIL::Const::to_bits() const
+{
+	if (auto bv = std::get_if<bitvectype>(&backing)) {
+		return *bv;
+	}
+	auto str = assert_get_str("Const::to_bits");
+	bitvectype b;
+	b.reserve(str.size() * 8);
+	for (int i = str.size()-1; i >= 0; i--) {
+		unsigned char ch = str[i];
+		for (int j = 0; j < 8; j++) {
+			b.push_back((ch & 1) != 0 ? State::S1 : State::S0);
+			ch = ch >> 1;
+		}
+	}
+	return b;
+}
+
+std::string RTLIL::Const::pretty_fmt() const {
```

Comment:
What are these new functions for?

Reply:
To move code in from elsewhere for deduplication and code reuse. This one is used in `Aig::Aig(Cell *cell)` now

Reply:
Right, previously, it was `name += stringf(":%d", p.second.as_int());` which would unpack string constants. We want to avoid that in all scenarios. Unpacking to bits is a safety measure, only because I can't guarantee it can't happen in some use case

Reply:
> which would unpack string constants. We want to avoid that in all scenarios.

Not here, these are the `WIDTH` parameters which are unlikely to have string backing in the first place

Reply:
We don't care about backing for parameters at all, at least not yet, they are almost never strings (the only exception I can think of is `FLAVOR` on `$check`; and parameters on user modules of course)

Reply:
Ah true, attributes are not covered by this. WIDTH parameters specifically still use `param` rather than `is_param` because their value is used. I still don't like enforcing non-bit-ness when we're only checking for param presence

Reply:
> WIDTH parameters specifically still use `param` rather than `is_param` because their value is used.

I was commenting on the cellaigs code which does `name += stringf(":%d", p.second.as_int());` which touches on the `WIDTH` params (`A_WIDTH` or `B_WIDTH`).

> I still don't like enforcing bitvec backing when we're only checking for param presence

Regarding the `is_param` vs `param` situation -- there aren't any gains to expect there, on the other hand any change we can shed off from this PR will make the review easier, which is important given the wide-reaching nature of it, for that reason I think it's better we drop that change.